File: probes/typescript/runtime-probe.mjs
#!/usr/bin/env node
// Read-only heuristics, not a sandbox or a compliance verdict.
import { createHash } from 'node:crypto';
import { isAbsolute, resolve } from 'node:path';
import { fileURLToPath, pathToFileURL } from 'node:url';
import { spawn } from 'node:child_process';
import { types } from 'node:util';

const VERSION = '1.1.0';
const SENSITIVE = /(?:api[_-]?key|auth|bearer|credential|passwd|password|private[_-]?key|secret|session|token)/i;
const PATTERNS = [/AKIA[0-9A-Z]{16}/, /gh[pousr]_[A-Za-z0-9_]{20,}/, /sk-[A-Za-z0-9_-]{20,}/, /eyJ[A-Za-z0-9_-]{8,}\.[A-Za-z0-9_-]{8,}\.[A-Za-z0-9_-]{8,}/, /-----BEGIN [A-Z ]*PRIVATE KEY-----/];
const HASH = /^sha256:[a-f0-9]{64}$/;
const TYPES = new Set(['null', 'array', 'undefined', 'object', 'function', 'boolean', 'number', 'bigint', 'string', 'symbol']);
const hash = (value) => `sha256:${createHash('sha256').update(value, 'utf8').digest('hex')}`;
const typeOf = (value) => value === null ? 'null' : Array.isArray(value) ? 'array' : typeof value;
const isObject = (value) => value !== null && (typeof value === 'object' || typeof value === 'function');

function report(target, scanned, truncated, fingerprints) {
  fingerprints.sort((a, b) => `${a.category}:${a.locationSha256}`.localeCompare(`${b.category}:${b.locationSha256}`));
  return {
    schemaVersion: 'canonical.runtime-probe/v1', runtime: 'typescript', probeVersion: VERSION, target,
    facts: { suspected_secret_count: fingerprints.length, scanned_binding_count: scanned, truncated }, fingerprints,
    limitations: [
      'Import runs customer code. Use only an explicitly authorized, externally isolated environment; a child process is not a security sandbox.',
      'Getters are not invoked, but Proxy descriptor traps can execute. Accessors, symbols, errors and depth/size limits make coverage incomplete.',
      'No inspected value is emitted by the observation serializer. No findings does not establish absence of secrets or compliance. Fingerprints remain sensitive evidence.',
    ],
  };
}

export function inspectBindings(target, bindings, maximum = 5000) {
  if (typeof target !== 'string' || target.length > 512 || /[\u0000-\u001f]/.test(target)) throw new TypeError('invalid target');
  if (!Number.isSafeInteger(maximum) || maximum < 1 || maximum > 10000 || !isObject(bindings)) throw new TypeError('invalid probe input');
  const queue = [], seen = new WeakSet(), fingerprints = [];
  let truncated = false, scanned = 0;
  function enqueue(value, path, depth) {
    let keys;
    try { keys = Reflect.ownKeys(value); } catch { truncated = true; return; }
    if (keys.some((key) => typeof key === 'symbol' && !(key === Symbol.toStringTag && types.isModuleNamespaceObject(value)))) truncated = true;
    keys = keys.filter((key) => typeof key === 'string');
    const limit = Math.min(256, maximum - queue.length);
    if (keys.length > limit) truncated = true;
    for (const name of keys.slice(0, limit).sort()) {
      let descriptor;
      try { descriptor = Object.getOwnPropertyDescriptor(value, name); } catch { truncated = true; continue; }
      if (!descriptor || !Object.hasOwn(descriptor, 'value')) { truncated = true; continue; }
      queue.push({ value: descriptor.value, path: [...path, name], name, depth });
    }
  }
  enqueue(bindings, [], 0);
  for (let cursor = 0; cursor < queue.length; cursor += 1) {
    const { value, path, name, depth } = queue[cursor];
    scanned += 1;
    const longString = typeof value === 'string' && value.length > 16384;
    if (longString) truncated = true;
    const valueMatch = typeof value === 'string' && !longString && PATTERNS.some((pattern) => pattern.test(value));
    if (SENSITIVE.test(name) || valueMatch) fingerprints.push({
      category: valueMatch ? 'secret_like_value' : 'sensitive_binding_name',
      locationSha256: hash(JSON.stringify(path)), valueSha256: valueMatch ? hash(value) : null, valueType: typeOf(value),
    });
    if (!isObject(value) || seen.has(value)) continue;
    seen.add(value);
    if (depth >= 4) { truncated = true; continue; }
    enqueue(value, path, depth + 1);
  }
  return report(target, scanned, truncated, fingerprints);
}

export function parseLimit(value, maximum) {
  if (typeof value !== 'string' || !/^[1-9][0-9]*$/.test(value)) throw new TypeError('invalid limit');
  const parsed = Number(value);
  if (!Number.isSafeInteger(parsed) || parsed > maximum) throw new TypeError('invalid limit');
  return parsed;
}

// Reconstruct subprocess observations from bounded, allowlisted metadata only.
// Imported code can write fd 3 itself; never reflect arbitrary child data.
function safeReport(value, target, maximum) {
  if (!value || !value.facts || !Array.isArray(value.fingerprints) || value.fingerprints.length > maximum) throw new Error('invalid observation');
  const { scanned_binding_count: scanned, truncated, suspected_secret_count: count } = value.facts;
  if (!Number.isSafeInteger(scanned) || scanned < 0 || scanned > maximum || typeof truncated !== 'boolean' || count !== value.fingerprints.length || count > scanned) throw new Error('invalid observation');
  const fingerprints = value.fingerprints.map((entry) => {
    if (!entry || !['secret_like_value', 'sensitive_binding_name'].includes(entry.category) || !(typeof entry.locationSha256 === 'string' && HASH.test(entry.locationSha256)) || !(entry.valueSha256 === null || (typeof entry.valueSha256 === 'string' && HASH.test(entry.valueSha256))) || !TYPES.has(entry.valueType)) throw new Error('invalid fingerprint');
    return { category: entry.category, locationSha256: entry.locationSha256, valueSha256: entry.valueSha256, valueType: entry.valueType };
  });
  return report(target, scanned, truncated, fingerprints);
}

export async function inspectModule(specifier, maximum = 5000, timeout = 5000) {
  parseLimit(String(maximum), 10000); parseLimit(String(timeout), 30000);
  if (typeof specifier !== 'string' || !specifier || specifier.length > 4096 || /[\u0000-\u001f]/.test(specifier)) throw new TypeError('invalid module');
  let url;
  if (specifier.startsWith('file:')) {
    url = new URL(specifier);
    if (url.search || url.hash || url.username || url.password) throw new TypeError('invalid module');
    fileURLToPath(url);
  } else {
    if (!isAbsolute(specifier) && !specifier.startsWith('./') && !specifier.startsWith('../')) throw new TypeError('local file required');
    url = pathToFileURL(resolve(specifier));
  }
  const target = hash(url.href);
  return new Promise((resolveResult, reject) => {
    const child = spawn(process.execPath, ['--max-old-space-size=128', fileURLToPath(new URL('./runtime-probe-worker.mjs', import.meta.url))], {
      stdio: ['ignore', 'pipe', 'pipe', 'pipe'],
      env: { ...process.env, NODE_OPTIONS: '', NODE_PATH: '', CANONICAL_PROBE_CHILD_MODULE: url.href, CANONICAL_PROBE_CHILD_MAXIMUM: String(maximum), CANONICAL_PROBE_CHILD_TARGET: target },
    });
    let settled = false, outputBytes = 0, resultBytes = 0;
    const chunks = [];
    const timer = setTimeout(() => finish(new Error('probe deadline exceeded')), timeout);
    function finish(error, result) {
      if (settled) return;
      settled = true; clearTimeout(timer); child.kill('SIGKILL');
      for (const stream of child.stdio) stream?.destroy();
      if (error) reject(error); else resolveResult(result);
    }
    for (const stream of [child.stdout, child.stderr]) stream.on('data', (chunk) => {
      outputBytes += chunk.length;
      if (outputBytes > 65536) finish(new Error('probe output limit exceeded'));
    });
    child.stdio[3].on('data', (chunk) => {
      resultBytes += chunk.length;
      if (resultBytes > 4194304) finish(new Error('probe result limit exceeded'));
      else chunks.push(chunk);
    });
    child.on('error', () => finish(new Error('probe process failed')));
    child.on('close', (code) => {
      if (settled) return;
      if (code !== 0) { finish(new Error('probe produced no observation')); return; }
      try { finish(null, safeReport(JSON.parse(Buffer.concat(chunks).toString('utf8')), target, maximum)); }
      catch { finish(new Error('invalid observation')); }
    });
  });
}

async function main() {
  if (process.argv.length !== 2) { process.stderr.write('runtime-probe: arguments are not supported; use documented environment keys\n'); return 2; }
  if (process.env.CANONICAL_PROBE_SELF_TEST === '1') {
    const value = `sk-${'a'.repeat(32)}`;
    const result = inspectBindings('self_test', { config: { token: value } }, 100);
    if (result.facts.suspected_secret_count !== 1 || JSON.stringify(result).includes(value)) return 1;
    process.stdout.write(`${JSON.stringify({ ok: true, probeVersion: VERSION })}\n`); return 0;
  }
  try {
    if (process.env.CANONICAL_PROBE_ALLOW_IMPORT !== '1') throw new Error('authorization required');
    const maximum = parseLimit(process.env.CANONICAL_PROBE_MAX_BINDINGS ?? '5000', 10000);
    const timeout = parseLimit(process.env.CANONICAL_PROBE_TIMEOUT_MS ?? '5000', 30000);
    const result = await inspectModule(process.env.CANONICAL_PROBE_MODULE, maximum, timeout);
    process.stdout.write(`${JSON.stringify(result, null, 2)}\n`);
    return result.facts.truncated ? 3 : 0;
  } catch {
    process.stderr.write('runtime-probe: no observation; check authorization, local module path, limits and isolation\n'); return 1;
  }
}
if (process.argv[1] && import.meta.url === pathToFileURL(resolve(process.argv[1])).href) process.exitCode = await main();

File: probes/python/runtime_probe.py
#!/usr/bin/env python3
"""Opt-in Python runtime exposure probe that never emits inspected values."""

from __future__ import annotations

import hashlib
import importlib
import json
import os
import re
import sys
from collections import deque
from collections.abc import Mapping
from dataclasses import dataclass
from types import ModuleType
from typing import Any

PROBE_VERSION = "1.0.0"
SCHEMA_VERSION = "canonical.runtime-probe/v1"
MODULE_PATTERN = re.compile(r"^[A-Za-z_][A-Za-z0-9_.]{0,159}$")
SENSITIVE_NAME = re.compile(
    r"(?:api[_-]?key|auth|bearer|credential|passwd|password|private[_-]?key|secret|session|token)",
    re.IGNORECASE,
)
SECRET_VALUE_PATTERNS = (
    re.compile(r"AKIA[0-9A-Z]{16}"),
    re.compile(r"gh[pousr]_[A-Za-z0-9_]{20,}"),
    re.compile(r"sk-[A-Za-z0-9_-]{20,}"),
    re.compile(r"eyJ[A-Za-z0-9_-]{8,}\.[A-Za-z0-9_-]{8,}\.[A-Za-z0-9_-]{8,}"),
    re.compile(r"-----BEGIN [A-Z ]*PRIVATE KEY-----"),
)


@dataclass(frozen=True)
class Fingerprint:
    """Non-reversible metadata for one possible exposure."""

    category: str
    location_sha256: str
    value_sha256: str | None
    value_type: str


@dataclass(frozen=True)
class ProbeResult:
    """Bounded probe output safe for later evidence normalization."""

    schema_version: str
    runtime: str
    probe_version: str
    target: str
    facts: dict[str, int | bool]
    fingerprints: tuple[Fingerprint, ...]
    limitations: tuple[str, ...]


def _digest(value: str) -> str:
    return "sha256:" + hashlib.sha256(value.encode("utf-8")).hexdigest()


def _looks_secret(value: str) -> bool:
    if len(value) > 16_384:
        return False
    return any(pattern.search(value) is not None for pattern in SECRET_VALUE_PATTERNS)


def inspect_bindings(target: str, bindings: Mapping[str, Any], maximum: int) -> ProbeResult:
    """Inspect inert containers and strings without invoking properties or callables."""

    queue: deque[tuple[str, Any, int]] = deque(
        (f"module.{name}", value, 0)
        for name, value in sorted(bindings.items())
        if not name.startswith("__")
    )
    seen: set[int] = set()
    fingerprints: list[Fingerprint] = []
    scanned = 0
    truncated = False

    while queue:
        if scanned >= maximum:
            truncated = True
            break
        location, value, depth = queue.popleft()
        scanned += 1
        identity = id(value)
        if identity in seen:
            continue
        seen.add(identity)

        leaf_name = location.rsplit(".", 1)[-1]
        name_match = SENSITIVE_NAME.search(leaf_name) is not None
        value_match = isinstance(value, str) and _looks_secret(value)
        if name_match or value_match:
            fingerprints.append(
                Fingerprint(
                    category="secret_like_value" if value_match else "sensitive_binding_name",
                    location_sha256=_digest(location),
                    value_sha256=_digest(value) if value_match else None,
                    value_type=type(value).__name__,
                )
            )

        if depth >= 4 or isinstance(value, (str, bytes, bytearray)):
            continue
        if type(value) is dict:
            for index, (key, nested) in enumerate(list(value.items())[:256]):
                queue.append((f"{location}.key:{index}", key, depth + 1))
                queue.append((f"{location}.value:{index}", nested, depth + 1))
        elif type(value) in (list, tuple):
            for index, nested in enumerate(value[:256]):
                queue.append((f"{location}.item:{index}", nested, depth + 1))

    fingerprints.sort(key=lambda item: (item.category, item.location_sha256))
    return ProbeResult(
        schema_version=SCHEMA_VERSION,
        runtime="python",
        probe_version=PROBE_VERSION,
        target=target,
        facts={
            "suspected_secret_count": len(fingerprints),
            "scanned_binding_count": scanned,
            "truncated": truncated,
        },
        fingerprints=tuple(fingerprints),
        limitations=(
            "Importing the target may execute its normal module initialization; run only in an authorized isolated environment.",
            "The probe inspects module globals and inert containers only; it does not invoke callables, descriptors, or properties.",
            "Fingerprints are heuristics and require human validation; no inspected value is emitted.",
        ),
    )


def inspect_module(module_name: str, maximum: int = 5_000) -> ProbeResult:
    """Explicitly import and inspect one allow-named customer module."""

    if MODULE_PATTERN.fullmatch(module_name) is None:
        raise ValueError("module name must be a dotted Python identifier")
    if not 1 <= maximum <= 10_000:
        raise ValueError("maximum must be between 1 and 10000")
    module: ModuleType = importlib.import_module(module_name)
    return inspect_bindings(module_name, vars(module), maximum)


def to_wire(result: ProbeResult) -> dict[str, Any]:
    """Project the internal immutable result into the shared camelCase JSON contract."""

    return {
        "schemaVersion": result.schema_version,
        "runtime": result.runtime,
        "probeVersion": result.probe_version,
        "target": result.target,
        "facts": result.facts,
        "fingerprints": [
            {
                "category": item.category,
                "locationSha256": item.location_sha256,
                "valueSha256": item.value_sha256,
                "valueType": item.value_type,
            }
            for item in result.fingerprints
        ],
        "limitations": list(result.limitations),
    }


def _self_test() -> None:
    secret = "sk-" + ("a" * 32)
    result = inspect_bindings("self_test", {"safe": 1, "api_token": secret}, 100)
    encoded = json.dumps(to_wire(result), sort_keys=True)
    assert result.facts["suspected_secret_count"] == 1
    assert secret not in encoded
    print(json.dumps({"ok": True, "probeVersion": PROBE_VERSION}, sort_keys=True))


def main() -> int:
    """Environment-only executable boundary; argv is intentionally unsupported."""

    if len(sys.argv) != 1:
        print("runtime-probe: command-line options are not supported; use documented environment keys", file=sys.stderr)
        return 2
    if os.environ.get("CANONICAL_PROBE_SELF_TEST") == "1":
        _self_test()
        return 0
    module_name = os.environ.get("CANONICAL_PROBE_MODULE")
    if module_name is None:
        print("runtime-probe: CANONICAL_PROBE_MODULE is required", file=sys.stderr)
        return 2
    maximum_text = os.environ.get("CANONICAL_PROBE_MAX_BINDINGS", "5000")
    try:
        maximum = int(maximum_text)
        result = inspect_module(module_name, maximum)
    except (ImportError, TypeError, ValueError) as error:
        print(f"runtime-probe: probe failed: {type(error).__name__}", file=sys.stderr)
        return 1
    print(json.dumps(to_wire(result), indent=2, sort_keys=True))
    return 0


if __name__ == "__main__":
    raise SystemExit(main())
